st2actions/st2actions/policies/retry.py

```diff
@@ -0,0 +1,141 @@
+# Licensed to the StackStorm, Inc ('StackStorm') under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import copy
+
+from st2common import log as logging
+from st2common.models.db.liveaction import LiveActionDB
+import st2common.services.action as action_services
+from st2common.constants.action import LIVEACTION_STATUS_FAILED
+from st2common.constants.action import LIVEACTION_STATUS_TIMED_OUT
+from st2common.util.enum import Enum
+from st2common.policies.base import ResourcePolicyApplicator
+
+__all__ = [
+    'RetryOnPolicy',
+    'ExecutionRetryPolicyApplicator'
+]
+
+LOG = logging.getLogger(__name__)
+
+VALID_RETRY_STATUSES = [
+    LIVEACTION_STATUS_FAILED,
+    LIVEACTION_STATUS_TIMED_OUT
+]
+
+# Maximum value for "max_retry_count" parameter
+MAXIMUM_RETRY_COUNT = 5
+
+
+class RetryOnPolicy(Enum):
+    FAILURE = 'failure'  # Retry on execution failure
+    TIMEOUT = 'timeout'  # Retry on execution timeout
+
+
+class ExecutionRetryPolicyApplicator(ResourcePolicyApplicator):
+    def __init__(self, policy_ref, policy_type, retry_on, max_retry_count=2):
+        """
+        :param retry_on: Condition to retry the execution on.
+        :type retry_on: ``str``
+
+        :param max_retry_count: Maximum number of times to try to retry an action.
+        :type max_retry_count: ``int``
+        """
+        # TODO: Also support sleep delay before retrying
```

Comment:
I guess difference between time based delay and concurrency based delay.  I feel that the current `DELAYED` state should probably be renamed to `BLOCKED` then there will be no confusion.